File: scraper-android/src/scraper.py
"""Google Play scraper wrapper using gplay-scraper package."""

import re
import logging
from datetime import date

import requests
from gplay_scraper import GPlayScraper

logger = logging.getLogger(__name__)


class AppNotFoundError(Exception):
    """Raised when gplay-scraper cannot find the requested app in this storefront."""

from .schemas import (
    AppIdentity,
    AppMetrics,
    ChartEntry,
    ChartResponse,
    DeveloperApp,
    DeveloperAppsResponse,
    LocalizedListingsResponse,
    Screenshot,
    SearchResponse,
    SearchResult,
    StoreListing,
)

ALL_APP_FIELDS = [
    "appId", "title", "summary", "description", "genre", "contentRating",
    "released", "lastUpdated", "version", "icon", "headerImage",
    "screenshots", "video", "installs", "realInstalls", "minInstalls",
    "score", "ratings", "histogram", "price", "free", "currency",
    "developer", "developerId", "developerEmail", "developerWebsite",
    "permissions", "dataSafety", "appUrl", "whatsNew",
]

SEARCH_FIELDS = ["appId", "title", "developer", "developerId", "icon", "score", "free", "price", "currency", "version"]

CHART_FIELDS = ["appId", "title", "icon", "developer", "developerId", "genre", "genreId", "price", "free", "score", "currency", "version"]

COLLECTION_MAP = {
    "top_free": "topselling_free",
    "top_paid": "topselling_paid",
    "top_grossing": "topgrossing",
}

DEVELOPER_FIELDS = ["appId", "title", "icon", "score", "ratings", "free", "genre", "genreId"]

scraper = GPlayScraper()


def _resolve_developer_id(info: dict) -> str | None:
    """Google Play developer page uses URL-encoded name, not numeric ID."""
    name = info.get("developer", "")
    if name:
        return name.replace(" ", "+")
    return info.get("developerId")


def fetch_identity(app_id: str, country: str = "us") -> AppIdentity:
    """Fetch app identity/details from Google Play."""
    info = scraper.app_get_fields(app_id, [
        "appId", "title", "developer", "developerId", "developerWebsite",
        "genre", "genreId", "released", "lastUpdated", "free",
        "version", "price", "currency",
    ], country=country)

    if not info:
        raise AppNotFoundError(f"App not found: {app_id} in {country}")

    version = info.get("version")
    if not version or version == "Varies with device":
        version = f"ag.{date.today().strftime('%Y%m%d')}"

    return AppIdentity(
        app_id=info.get("appId", app_id),
        name=info.get("title", ""),
        publisher_name=info.get("developer", ""),
        publisher_external_id=_resolve_developer_id(info),
        publisher_url=info.get("developerWebsite"),
        category=info.get("genre", ""),
        category_id=info.get("genreId"),
        supported_locales=None,
        original_release_date=info.get("released"),
        is_free=info.get("free", True),
        price=info.get("price", 0) or 0,
        currency=info.get("currency"),
        version=version,
        current_version_release_date=info.get("lastUpdated") or date.today().isoformat(),
    )


def fetch_listing(app_id: str, locale: str = "en", country: str = "us") -> StoreListing:
    """Fetch store listing for a specific locale."""
    info = scraper.app_get_fields(app_id, [
        "title", "summary", "description", "whatsNew", "icon",
        "screenshots", "video", "price", "free", "currency",
    ], lang=locale, country=country)

    if not info:
        raise AppNotFoundError(f"App not found: {app_id} in {country}/{locale}")

    screenshots = []
    for i, url in enumerate(info.get("screenshots", []) or []):
        screenshots.append(Screenshot(url=url, device_type="phone", order=i))

    description = info.get("description", "") or ""

    whats_new = info.get("whatsNew")
    if isinstance(whats_new, list):
        whats_new = "\n".join(whats_new)
    if not whats_new:
        whats_new = None

    return StoreListing(
        platform="android",
        locale=locale,
        title=info.get("title", ""),
        subtitle=info.get("summary"),
        description=description,
        promotional_text=None,
        whats_new=whats_new,
        icon_url=info.get("icon"),
        screenshots=screenshots,
        video_url=info.get("video"),
        price=info.get("price", 0) or 0,
        currency=info.get("currency"),
        description_length=len(description),
    )


def fetch_localized_listings(
    app_id: str, locales: list[str]
) -> LocalizedListingsResponse:
    """Fetch store listings for multiple locales."""
    listings = []

    for locale in locales:
        try:
            listing = fetch_listing(app_id, locale=locale)
            listings.append(listing)
        except Exception as exc:
            logger.warning(
                "fetch_localized_listings: locale failed",
                extra={"app_id": app_id, "locale": locale, "reason": str(exc)},
            )
            continue

    return LocalizedListingsResponse(listings=listings)


def fetch_metrics(app_id: str, country: str = "us") -> AppMetrics:
    """Fetch app metrics (rating, installs, etc.)."""
    info = scraper.app_get_fields(app_id, [
        "score", "ratings", "histogram", "realInstalls", "minInstalls",
    ], country=country)

    if not info:
        raise AppNotFoundError(f"App not found: {app_id} in {country}")

    installs = info.get("realInstalls") or info.get("minInstalls")
    installs_range = f"{installs:,}+" if installs else None

    histogram = info.get("histogram")
    rating_breakdown = None
    if histogram and len(histogram) == 5:
        rating_breakdown = {
            "5": histogram[4],
            "4": histogram[3],
            "3": histogram[2],
            "2": histogram[1],
            "1": histogram[0],
        }
    elif (info.get("ratings") or 0) > 0:
        logger.warning(
            "fetch_metrics: rating_breakdown missing despite non-zero ratings",
            extra={
                "app_id": app_id,
                "country": country,
                "rating_count": info.get("ratings"),
                "histogram_len": len(histogram) if histogram else 0,
            },
        )

    return AppMetrics(
        rating=info.get("score", 0) or 0,
        rating_count=info.get("ratings", 0) or 0,
        rating_breakdown=rating_breakdown,
        installs_range=installs_range,
    )


def _scrape_developer_page(developer_id: str) -> list[str]:
    """Scrape Google Play developer page to get app IDs."""
    url = f"https://play.google.com/store/apps/developer?id={developer_id}&hl=en"
    headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)"}

    try:
        resp = requests.get(url, headers=headers, timeout=15)
        if resp.status_code != 200:
            logger.warning(
                "_scrape_developer_page: non-200",
                extra={"developer_id": developer_id, "status": resp.status_code},
            )
            return []
        app_ids = re.findall(r"details\?id=([a-zA-Z0-9._]+)", resp.text)
        return list(dict.fromkeys(app_ids))
    except Exception as e:
        logger.warning(
            "_scrape_developer_page: exception",
            extra={"developer_id": developer_id, "reason": str(e)},
        )
        return []


def fetch_developer_apps(developer_id: str) -> DeveloperAppsResponse:
    """Fetch all apps by a developer via developer page scraping + enrichment."""
    app_ids = _scrape_developer_page(developer_id)

    apps = []
    for app_id in app_ids:
        try:
            info = scraper.app_get_fields(app_id, DEVELOPER_FIELDS + ["free"])
            if not info:
                logger.warning(
                    "fetch_developer_apps: empty info",
                    extra={"developer_id": developer_id, "app_id": app_id},
                )
                continue
            apps.append(
                DeveloperApp(
                    external_id=info.get("appId", app_id),
                    name=info.get("title", ""),
                    icon_url=info.get("icon"),
                    rating=info.get("score"),
                    rating_count=info.get("ratings"),
                    price_model="free" if info.get("free", True) else "paid",
        price=info.get("price", 0) or 0,
        currency=info.get("currency"),
                    category=info.get("genre"),
                    category_id=info.get("genreId"),
                )
            )
        except Exception as exc:
            logger.warning(
                "fetch_developer_apps: enrichment failed",
                extra={
                    "developer_id": developer_id,
                    "app_id": app_id,
                    "reason": str(exc),
                },
            )
            continue

    return DeveloperAppsResponse(apps=apps)


def fetch_chart(
    collection: str = "top_free",
    category: str = "APPLICATION",
    country: str = "us",
    count: int = 100,
) -> ChartResponse:
    """Fetch top chart rankings from Google Play."""
    collection_value = COLLECTION_MAP.get(collection)
    if not collection_value:
        raise ValueError(f"Invalid collection: {collection}. Use: {', '.join(COLLECTION_MAP.keys())}")

    results = scraper.list_get_fields(
        collection=collection_value,
        fields=CHART_FIELDS,
        category=category,
        count=count,
        lang="en",
        country=country,
    )

    entries = []
    for i, info in enumerate(results or []):
        entries.append(
            ChartEntry(
                rank=i + 1,
                app_id=info.get("appId", ""),
                name=info.get("title", ""),
                icon_url=info.get("icon"),
                developer=info.get("developer"),
                developer_id=_resolve_developer_id(info),
                genre=info.get("genre"),
                genre_id=info.get("genreId"),
                price=info.get("price", 0),
                free=info.get("free", True),
                rating=info.get("score"),
                version=info.get("version"),
                currency=info.get("currency"),
            )
        )

    return ChartResponse(results=entries)


def search_apps(term: str, limit: int = 10, country: str = "us") -> SearchResponse:
    """Search for apps on Google Play."""
    results = scraper.search_get_fields(term, SEARCH_FIELDS, country=country)

    mapped = []
    for info in (results or [])[:limit]:
        mapped.append(
            SearchResult(
                app_id=info.get("appId", ""),
                name=info.get("title", ""),
                developer=info.get("developer", ""),
                developer_id=_resolve_developer_id(info),
                icon_url=info.get("icon"),
                rating=info.get("score"),
                version=info.get("version"),
                price=info.get("price", 0),
                free=info.get("free", True),
                currency=info.get("currency"),
            )
        )

    return SearchResponse(results=mapped)

File: scraper-ios/src/scraper.ts
/**
 * App Store scraper wrapper using app-store-scraper package + web scraping fallback.
 */

// @ts-expect-error - app-store-scraper has no type definitions
import store from "app-store-scraper";
import type {
  AppIdentity,
  AppMetrics,
  AppReview,
  ChartEntry,
  DeveloperApp,
  Screenshot,
  SearchResult,
  StoreListing,
} from "./schemas.js";

/**
 * Scrape App Store web page for data not available via the scraper package
 * (subtitle, screenshots, video, rating breakdown).
 */
async function scrapeAppStorePage(
  trackId: string,
  country: string = "us"
): Promise<{
  subtitle: string | null;
  screenshots: Screenshot[];
  video_url: string | null;
  rating_breakdown: Record<string, number> | null;
}> {
  const result = {
    subtitle: null as string | null,
    screenshots: [] as Screenshot[],
    video_url: null as string | null,
    rating_breakdown: null as Record<string, number> | null,
  };

  try {
    const url = `https://apps.apple.com/${country}/app/id${trackId}`;
    const response = await fetch(url, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36",
        "Accept-Language": "en-US,en;q=0.9",
      },
    });

    if (!response.ok) return result;

    const html = await response.text();
    const match = html.match(
      /<script[^>]*id="serialized-server-data"[^>]*>(.+?)<\/script>/s
    );
    if (!match) return result;

    const serverData = JSON.parse(match[1].trim());
    if (!serverData) return result;

    const appData =
      serverData?.data?.[0]?.data ?? serverData?.[0]?.data ?? {};
    if (!appData || Object.keys(appData).length === 0) return result;

    // Subtitle
    result.subtitle = appData?.lockup?.subtitle ?? null;

    const mapping = appData?.shelfMapping ?? {};
    let order = 0;

    // Phone screenshots
    for (const item of mapping?.["product_media_phone_"]?.items ?? []) {
      if (item?.screenshot?.template) {
        const screenshotUrl = item.screenshot.template
          .replace("{w}", "1290")
          .replace("{h}", "2796")
          .replace("{c}", "bb")
          .replace("{f}", "jpg");
        result.screenshots.push({
          url: screenshotUrl,
          device_type: "iphone",
          order: order++,
        });
      }
      if (item?.video?.videoUrl && !result.video_url) {
        result.video_url = item.video.videoUrl;
      }
    }

    // iPad screenshots
    for (const item of mapping?.["product_media_pad_"]?.items ?? []) {
      if (item?.screenshot?.template) {
        const screenshotUrl = item.screenshot.template
          .replace("{w}", "2048")
          .replace("{h}", "2732")
          .replace("{c}", "bb")
          .replace("{f}", "jpg");
        result.screenshots.push({
          url: screenshotUrl,
          device_type: "ipad",
          order: order++,
        });
      }
    }

    // Rating breakdown
    const ratingsItems = mapping?.productRatings?.items ?? [];
    if (ratingsItems.length > 0) {
      const counts = ratingsItems[0]?.ratingCounts;
      if (counts && counts.length === 5) {
        result.rating_breakdown = {
          "5": counts[0],
          "4": counts[1],
          "3": counts[2],
          "2": counts[3],
          "1": counts[4],
        };
      }
    }
  } catch {
    // Silently fail — web scraping is a best-effort fallback
  }

  return result;
}

export async function fetchIdentity(appId: string, country: string = "us", lang?: string): Promise<AppIdentity> {
  const opts: Record<string, any> = { id: Number(appId), country };
  if (lang) opts.lang = lang;
  const info = await store.app(opts);

  return {
    app_id: String(info.id),
    name: info.title || "",
    publisher_name: info.developer || "",
    publisher_external_id: info.developerId ? String(info.developerId).split("?")[0] : null,
    publisher_url: info.developerUrl || null,
    category: info.primaryGenre || "",
    category_id: info.primaryGenreId != null ? String(info.primaryGenreId) : null,
    content_rating: info.contentRating || null,
    supported_locales: info.languages || null,
    original_release_date: info.released
      ? new Date(info.released).toISOString().slice(0, 10)
      : null,
    price_model: info.free ? "free" : "paid",
    price: info.price ?? 0,
    currency: info.currency ?? null,
    store_url: info.url || null,
    version: info.version || `ag.${new Date().toISOString().slice(0, 10).replace(/-/g, '')}`,
    current_version_release_date: info.currentVersionReleaseDate
      ? new Date(info.currentVersionReleaseDate).toISOString().slice(0, 10)
      : new Date().toISOString().slice(0, 10),
  };
}

export async function fetchListing(
  appId: string,
  country: string = "us",
  lang?: string
): Promise<StoreListing> {
  const opts: Record<string, any> = { id: Number(appId), country };
  if (lang) opts.lang = lang;
  const [info, webData] = await Promise.all([
    store.app(opts),
    scrapeAppStorePage(appId, country),
  ]);

  const screenshots: Screenshot[] = [];
  let order = 0;

  const iphoneUrls = info.screenshots || [];
  for (const url of iphoneUrls) {
    screenshots.push({ url, device_type: "iphone", order: order++ });
  }

  // If package returned no iPhone screenshots, use web-scraped ones
  if (iphoneUrls.length === 0) {
    for (const s of webData.screenshots.filter((s: { device_type: string }) => s.device_type === "iphone")) {
      screenshots.push({ ...s, order: order++ });
    }
  }

  for (const url of info.ipadScreenshots || []) {
    screenshots.push({ url, device_type: "ipad", order: order++ });
  }

  const finalScreenshots = screenshots;

  const description = info.description || "";

  return {
    platform: "ios",
    locale: lang || country,
    title: info.title || "",
    subtitle: info.subtitle || webData.subtitle || null,
    short_description: null,
    description,
    promotional_text: info.promotionalText || null,
    whats_new: info.releaseNotes || "Bug fixes and performance improvements.",
    icon_url: info.icon || null,
    screenshots: finalScreenshots,
    video_url: info.videoUrl || webData.video_url || null,
    description_length: description.length,
    price: info.price ?? 0,
    currency: info.currency ?? null,
  };
}

export async function fetchLocalizedListings(
  appId: string,
  countries: string[]
): Promise<StoreListing[]> {
  const listings: StoreListing[] = [];

  for (const country of countries) {
    try {
      const listing = await fetchListing(appId, country);
      listings.push(listing);
    } catch {
      continue;
    }
  }

  return listings;
}

export async function fetchMetrics(appId: string, country: string = "us", lang?: string): Promise<AppMetrics> {
  const opts: Record<string, any> = { id: Number(appId), country };
  if (lang) opts.lang = lang;
  const [info, webData] = await Promise.all([
    store.app(opts),
    scrapeAppStorePage(appId, country),
  ]);

  let ratingBreakdown: Record<string, number> | null = null;
  if (info.histogram) {
    ratingBreakdown = {
      "5": info.histogram["5"] || 0,
      "4": info.histogram["4"] || 0,
      "3": info.histogram["3"] || 0,
      "2": info.histogram["2"] || 0,
      "1": info.histogram["1"] || 0,
    };
  }

  // Use web-scraped breakdown if package didn't provide one
  if (!ratingBreakdown && webData.rating_breakdown) {
    ratingBreakdown = webData.rating_breakdown;
  }

  return {
    rating: info.score || 0,
    rating_count: info.reviews || 0,
    rating_breakdown: ratingBreakdown,
    installs_range: null,
    file_size_bytes: info.size ? Number(info.size) : null,
  };
}

export async function fetchReviews(
  appId: string,
  country: string = "us",
  page: number = 1
): Promise<{
  reviews: AppReview[];
  rating_breakdown: Record<string, number> | null;
}> {
  const reviewData = await store.reviews({
    id: Number(appId),
    country,
    page,
    sort: store.sort.RECENT,
  });

  const reviews: AppReview[] = reviewData.map((r: any) => ({
    external_id: String(r.id),
    author: r.userName || null,
    title: r.title || null,
    body: r.text || null,
    rating: r.score || 0,
    review_date: r.date ? new Date(r.date).toISOString().slice(0, 10) : null,
    app_version: r.version || null,
    country_code: country.toUpperCase(),
  }));

  return { reviews, rating_breakdown: null };
}

export async function fetchDeveloperApps(
  developerId: string
): Promise<DeveloperApp[]> {
  const apps = await store.developer({ devId: Number(developerId) });

  return apps.map((info: any) => ({
    external_id: String(info.id),
    name: info.title || "",
    icon_url: info.icon || null,
    rating: info.score || null,
    rating_count: info.reviews || null,
    price_model: info.free ? "free" : "paid",
    price: info.price ?? 0,
    currency: info.currency ?? null,
    category: info.primaryGenre || null,
    category_id: info.primaryGenreId != null ? String(info.primaryGenreId) : null,
  }));
}

const COLLECTION_MAP: Record<string, string> = {
  top_free: store.collection.TOP_FREE_IOS,
  top_paid: store.collection.TOP_PAID_IOS,
  top_grossing: store.collection.TOP_GROSSING_IOS,
};

export async function fetchChart(
  collection: string,
  category?: number,
  country: string = "us",
  num: number = 200
): Promise<ChartEntry[]> {
  const collectionValue = COLLECTION_MAP[collection];
  if (!collectionValue) {
    throw new Error(`Invalid collection: ${collection}. Use: ${Object.keys(COLLECTION_MAP).join(", ")}`);
  }

  const safeNum = Math.max(num, 2);
  const opts: Record<string, any> = { collection: collectionValue, num: safeNum, country };
  if (category) opts.category = category;

  let results;
  try {
    results = await store.list(opts);
  } catch (err: any) {
    console.warn(`[chart] empty or failed: ${collection} ${country} cat=${category ?? 'all'} — ${err.message}`);
    return [];
  }

  if (!results || !Array.isArray(results)) {
    return [];
  }

  return results.map((app: any, index: number) => ({
    rank: index + 1,
    app_id: String(app.id),
    name: app.title || "",
    icon_url: app.icon || null,
    developer: app.developer || null,
    developer_id: app.developerId ? String(app.developerId).split("?")[0] : null,
    genre: app.genre || null,
    genre_id: app.genreId ? String(app.genreId) : null,
    price: app.price ?? null,
    currency: app.currency ?? null,
    free: app.free ?? true,
    rating: null,
    version: app.version || null,
    released: app.released ? new Date(app.released).toISOString().slice(0, 10) : null,
  }));
}

export async function searchApps(
  term: string,
  num: number = 10,
  country: string = "us"
): Promise<SearchResult[]> {
  const results = await store.search({ term, num, country });

  return results.map((info: any) => ({
    app_id: String(info.id),
    name: info.title || "",
    developer: info.developer || "",
    developer_id: info.developerId ? String(info.developerId).split("?")[0] : null,
    icon_url: info.icon || null,
    rating: info.score || null,
    version: info.version || null,
    price: info.price || null,
    currency: info.currency ?? null,
    free: info.free ?? true,
  }));
}
